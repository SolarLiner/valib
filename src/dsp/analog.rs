use nalgebra::Complex;
use crate::dsp::DSP;

pub trait DspAnalog<const I: usize, const O: usize>: DSP<I, O> {
    fn h_s(&self, s: [Complex<Self::Sample>; I]) -> [Complex<Self::Sample>; O];
}
