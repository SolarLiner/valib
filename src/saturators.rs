use num_traits::NumCast;
use numeric_literals::replace_float_literals;

use crate::Scalar;

pub trait Saturator<T: Scalar>: Default {
    /// Saturate an input with a frozen state.
    fn saturate(&self, x: T) -> T;

    /// Update the state given an input.
    #[inline(always)]
    fn update_state(&mut self, _x: T) {}

    #[inline(always)]
    #[replace_float_literals(T::from(literal).unwrap())]
    fn sat_diff(&self, x: T) -> T {
        (self.saturate(x + 1e-4) - self.saturate(x)) / 1e-4
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Linear;

impl<S: Scalar> Saturator<S> for Linear {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x
    }

    #[inline(always)]
    fn sat_diff(&self, _: S) -> S {
        S::one()
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Tanh;

impl<S: Scalar> Saturator<S> for Tanh {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x.tanh()
    }

    #[inline(always)]
    #[replace_float_literals(S::from(literal).unwrap())]
    fn sat_diff(&self, x: S) -> S {
        1. - x.tanh().powi(2)
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Clipper;

impl<S: Scalar> Saturator<S> for Clipper {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x.min(S::one()).max(S::one().neg())
    }

    #[inline(always)]
    #[replace_float_literals(S::from(literal).unwrap())]
    fn sat_diff(&self, x: S) -> S {
        if x < -1. || x > 1. {
            0.
        } else {
            1.
        }
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct DiodeClipper<T> {
    param: T,
}

impl<T: Scalar> Default for DiodeClipper<T> {
    fn default() -> Self {
        // Parameter fit from a KiCad simulation of a buffer -> 2x3 diode clipper stage from -12V to 12V,
        // which yields a maximum difference of -20 dB.
        Self {
            param: T::from(0.75975449).unwrap(),
        }
    }
}

impl<T: Scalar> Saturator<T> for DiodeClipper<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        self.param * x / (self.param + x.abs())
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        let frac = self.param / (self.param + x.abs());
        frac.powi(2)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Blend<T, S> {
    amt: T,
    inner: S,
}

impl<T: Scalar, S: Saturator<T>> Saturator<T> for Blend<T, S> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        x + self.amt * (self.inner.saturate(x) - x)
    }

    #[inline(always)]
    fn update_state(&mut self, x: T) {
        self.inner.update_state(x)
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        T::one() + self.amt * (self.inner.sat_diff(x) - T::one())
    }
}

impl<T: NumCast, S: Default> Default for Blend<T, S> {
    fn default() -> Self {
        Self {
            amt: T::from(0.5).unwrap(),
            inner: S::default(),
        }
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Dynamic<T> {
    Linear,
    Tanh,
    HardClipper,
    DiodeClipper(DiodeClipper<T>),
    SoftClipper(Blend<T, DiodeClipper<T>>),
}

impl<T: Scalar> Saturator<T> for Dynamic<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.saturate(x),
            Self::HardClipper => Clipper.saturate(x),
            Self::Tanh => Tanh.saturate(x),
            Self::DiodeClipper(clip) => clip.saturate(x),
            Self::SoftClipper(clip) => clip.saturate(x),
        }
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.sat_diff(x),
            Self::HardClipper => Clipper.sat_diff(x),
            Self::Tanh => Tanh.sat_diff(x),
            Self::DiodeClipper(clip) => clip.sat_diff(x),
            Self::SoftClipper(clip) => clip.sat_diff(x),
        }
    }
}

impl<T> Default for Dynamic<T> {
    fn default() -> Self {
        Self::Linear
    }
}
