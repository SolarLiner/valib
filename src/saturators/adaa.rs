use num_traits::Zero;
use numeric_literals::replace_float_literals;
use simba::simd::SimdBool;

use crate::dsp::DSP;
use crate::saturators::{Asinh, Blend, Clipper, Saturator, Tanh};
use crate::Scalar;

/// Trait for functions that have antiderivatives.
pub trait Antiderivative<T> {
    fn evaluate(&self, x: T) -> T;

    fn antiderivative(&self, x: T) -> T;
}

pub trait Antiderivative2<T>: Antiderivative<T> {
    fn antiderivative2(&self, x: T) -> T;
}

impl<T: Scalar> Antiderivative<T> for Tanh {
    fn evaluate(&self, x: T) -> T {
        x.simd_tanh()
    }

    fn antiderivative(&self, x: T) -> T {
        x.simd_cosh().simd_ln()
    }
}

impl<T: Scalar> Antiderivative<T> for Asinh {
    fn evaluate(&self, x: T) -> T {
        x.simd_asinh()
    }

    fn antiderivative(&self, x: T) -> T {
        let x0 = x * x + T::one();
        x * x.simd_asinh() - x0.simd_sqrt()
    }
}

impl<T: Scalar> Antiderivative2<T> for Asinh {
    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative2(&self, x: T) -> T {
        let x0 = x * x + 1.0;
        let x1 = 2.0 * x * x - 1.0;
        let x2 = -x * 3.0 * x0.simd_sqrt() / 4.0;
        let x3 = x1 * x.simd_asinh() / 4.0;
        x2 + x3
    }
}

impl<T: Scalar> Antiderivative<T> for Clipper {
    #[replace_float_literals(T::from_f64(literal))]
    fn evaluate(&self, x: T) -> T {
        x.simd_clamp(-1.0, 1.0)
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative(&self, x: T) -> T {
        let lower = -x;
        let upper = x;
        let middle = (0.5) * (x * x + T::one());
        let is_lower = x.simd_lt(-T::one());
        let is_higher = x.simd_gt(T::one());
        lower.select(is_lower, upper.select(is_higher, middle))
    }
}

impl<T: Scalar> Antiderivative2<T> for Clipper {
    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative2(&self, x: T) -> T {
        let lower = -0.5 * x * (x - 2.0);
        let upper = x * x / 2.0 * x * T::simd_recip(3.0);
        let middle = T::simd_recip(6.0) * (x * x * x + 9.0 * x + 1.0);
        let is_lower = x.simd_lt(-1.0);
        let is_higher = x.simd_gt(1.0);
        lower.select(is_lower, upper.select(is_higher, middle))
    }
}

impl<T: Scalar, S: Antiderivative<T>> Antiderivative<T> for Blend<T, S> {
    fn evaluate(&self, x: T) -> T {
        x + (self.inner.evaluate(x) - x) * self.amt
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative(&self, x: T) -> T {
        let ad_x = 0.5 * x * x;
        self.inner.antiderivative(x) * self.amt + ad_x * (1.0 + self.amt)
    }
}

impl<T: Scalar, S: Antiderivative2<T>> Antiderivative2<T> for Blend<T, S> {
    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative2(&self, x: T) -> T {
        let ad_x = x + x * x * x / 6.0;
        self.inner.antiderivative2(x) * self.amt + ad_x * (1.0 + self.amt)
    }
}

#[derive(Debug, Copy, Clone)]
pub struct Adaa<T, S, const ORDER: usize> {
    pub epsilon: T,
    pub inner: S,
    memory: [T; ORDER],
}

impl<T: Scalar, S, const ORDER: usize> Adaa<T, S, ORDER> {
    pub fn new(inner: S) -> Self {
        Self {
            epsilon: T::from_f64(1e-3),
            inner,
            memory: [T::zero(); ORDER],
        }
    }
}

impl<T: Scalar, S: Default, const ORDER: usize> Default for Adaa<T, S, ORDER> {
    fn default() -> Self {
        Self::new(S::default())
    }
}

impl<T: Scalar, S: Antiderivative<T>> Adaa<T, S, 1> {
    #[replace_float_literals(T::from_f64(literal))]
    pub fn next_sample_immutable(&self, x: T) -> T {
        let den = x - self.memory[0];
        let below = den.simd_abs().simd_lt(self.epsilon);
        below.if_else(
            || self.inner.evaluate((x + self.memory[0]) / 2.0),
            || {
                let num = self.inner.antiderivative(x) - self.inner.antiderivative(self.memory[0]);
                num / den
            },
        )
    }

    pub fn commit_sample(&mut self, x: T) {
        self.memory = [x];
    }

    pub fn next_sample(&mut self, x: T) -> T {
        let y = self.next_sample_immutable(x);
        self.commit_sample(x);
        y
    }
}

impl<T: Scalar, S: Antiderivative<T> + Saturator<T>> Saturator<T> for Adaa<T, S, 1> {
    fn saturate(&self, x: T) -> T {
        self.next_sample_immutable(x)
    }

    fn update_state(&mut self, x: T, y: T) {
        self.commit_sample(x);
        self.inner.update_state(x, y);
    }

    fn sat_diff(&self, x: T) -> T {
        self.inner.sat_diff(x)
    }
}

impl<T: Scalar, S: Antiderivative<T> + DSP<1, 1, Sample = T>> DSP<1, 1> for Adaa<T, S, 1> {
    type Sample = T;

    fn process(&mut self, [x]: [Self::Sample; 1]) -> [Self::Sample; 1] {
        [self.next_sample(x)]
    }
}

impl<T: Scalar, S: Antiderivative2<T> + Saturator<T>> Saturator<T> for Adaa<T, S, 2> {
    #[replace_float_literals(T::from_f64(literal))]
    fn saturate(&self, x: T) -> T {
        let [x1, x2] = self.memory;
        let den1 = x - x1;
        let den2 = x1 - x2;
        let den3 = x - x2;
        let below1 = den1.simd_abs().simd_lt(self.epsilon);
        let below2 = den2.simd_abs().simd_lt(self.epsilon);
        let below3 = den3.simd_abs().simd_lt(self.epsilon);
        (below1 | below2 | below3).if_else(
            || self.inner.evaluate((x + x1) / 2.0),
            || {
                let num1 = self.inner.antiderivative(x) - self.inner.antiderivative2(x1);
                let num2 = self.inner.antiderivative2(x1) - self.inner.antiderivative2(x2);
                den3.simd_recip() * (num1 / den1 + num2 / den2)
            },
        )
    }

    fn update_state(&mut self, x: T, y: T) {
        self.inner.update_state(x, y);
        self.memory.swap(0, 1);
        self.memory[0] = x;
    }

    fn sat_diff(&self, x: T) -> T {
        self.inner.sat_diff(x)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rstest::rstest;
    use std::f64::consts::TAU;

    #[rstest]
    #[case("tanh", Adaa::<_, Tanh, 1>::default())]
    #[case("asinh", Adaa::<_, Asinh, 1>::default())]
    #[case("clipper", Adaa::<_, Clipper, 1>::default())]
    fn test_adaa1<S: Antiderivative<f64> + Saturator<f64>>(
        #[case] name: &str,
        #[case] mut adaa: Adaa<f64, S, 1>,
    ) {
        let samplerate = 100.0;
        let f = 10.0;
        let input: [_; 100] =
            std::array::from_fn(|i| i as f64 / samplerate).map(|t| 3.0 * f64::sin(TAU * f * t));
        let output = input.map(|x| {
            let y = adaa.saturate(x);
            adaa.update_state(x, y);
            y
        });

        let name = format!("test_adaa1_{name}");
        insta::assert_csv_snapshot!(name, &output as &[_], { "[]" => insta::rounded_redaction(3) })
    }

    #[rstest]
    #[case("clipper", Adaa::<_, Clipper, 2>::default())]
    #[case("asinh", Adaa::<_, Asinh, 2>::default())]
    fn test_adaa2<S: Antiderivative2<f64> + Saturator<f64>>(
        #[case] name: &str,
        #[case] mut adaa: Adaa<f64, S, 2>,
    ) {
        let samplerate = 100.0;
        let f = 10.0;
        let input: [_; 100] =
            std::array::from_fn(|i| i as f64 / samplerate).map(|t| 3.0 * f64::sin(TAU * f * t));
        let output = input.map(|x| {
            let y = adaa.saturate(x);
            adaa.update_state(x, y);
            y
        });

        let name = format!("test_adaa2_{name}",);
        insta::assert_csv_snapshot!(name, &output as &[_], { "[]" => insta::rounded_redaction(3) })
    }
}
