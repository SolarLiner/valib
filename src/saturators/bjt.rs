use crate::dsp::{DSPMeta, DSPProcess};
use crate::math::smooth_clamp;
use crate::saturators::Saturator;
use crate::Scalar;
use numeric_literals::replace_float_literals;

/// Memoryless nonlinearity for a BJT NPN Transistor in a common-collector configuration.
/// `xbias` and `ybias` are empirical values that can be fit to DC transfer data or to recenter the
/// signal before and after the saturation.
#[derive(Debug, Copy, Clone)]
pub struct CommonCollector<T> {
    pub vcc: T,
    pub vee: T,
    pub xbias: T,
    pub ybias: T,
}

impl<T: Scalar> Default for CommonCollector<T> {
    fn default() -> Self {
        Self {
            vcc: T::from_f64(4.5),
            vee: T::from_f64(-4.5),
            xbias: T::from_f64(0.77),
            ybias: T::from_f64(-0.77),
        }
    }
}

#[profiling::all_functions]
impl<T: Scalar> Saturator<T> for CommonCollector<T> {
    #[replace_float_literals(T::from_f64(literal))]
    fn saturate(&self, x: T) -> T {
        smooth_clamp(0.1, x + self.xbias, self.vee, self.vcc) + self.ybias
    }
}

impl<T: Scalar> DSPMeta for CommonCollector<T> {
    type Sample = T;
}

#[profiling::all_functions]
impl<T: Scalar> DSPProcess<1, 1> for CommonCollector<T> {
    fn process(&mut self, [x]: [Self::Sample; 1]) -> [Self::Sample; 1] {
        [self.saturate(x)]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::util::lerp;
    use crate::util::tests::{Plot, Series};
    use plotters::prelude::BLUE;

    #[test]
    fn test_common_collector() {
        const N: usize = 100;
        const STEP: f32 = 1. / N as f32;
        let sat = CommonCollector {
            vee: -4.5,
            vcc: 4.5,
            xbias: 0.770,
            ybias: -0.770,
        };
        let input: [f32; N] = std::array::from_fn(|i| lerp(i as f32 * STEP, -10., 10.));
        let output = input.map(|x| sat.saturate(x));
        Plot {
            title: "Common Collector saturator",
            bode: false,
            series: &[Series {
                label: "Output",
                samplerate: N as f32,
                color: &BLUE,
                series: &output,
            }],
        }
        .create_svg("plots/saturators/bjt/common_collector.svg");
        insta::assert_csv_snapshot!(&output as &[_], { "[]" => insta::rounded_redaction(4)});
    }
}
