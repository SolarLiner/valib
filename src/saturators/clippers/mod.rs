use nalgebra::{SMatrix, SVector};
use num_traits::Float;
use numeric_literals::replace_float_literals;
use simba::simd::SimdBool;

use super::adaa::Antiderivative;
use crate::dsp::DSPMeta;
use crate::dsp::DSPProcess;
use crate::math::nr::{newton_rhapson_tol_max_iter, RootEq};
use crate::saturators::MultiSaturator;
use crate::{saturators::Saturator, Scalar};

mod diode_clipper_model_data;

#[derive(Debug, Copy, Clone)]
pub struct DiodeClipper<T> {
    pub isat: T,
    pub n: T,
    pub vt: T,
    pub vin: T,
    pub num_diodes_fwd: T,
    pub num_diodes_bwd: T,
    pub sim_tol: T,
    pub max_iter: usize,
    last_vout: T,
}

impl<T: Copy> DiodeClipper<T> {
    pub fn last_output(&self) -> T {
        self.last_vout
    }
}

impl<T: Copy> DiodeClipper<T> {
    pub fn reset(&mut self) {
        self.last_vout = self.vin;
    }
}

impl<T: Scalar> RootEq<T, 1> for DiodeClipper<T> {
    #[cfg_attr(not(test), inline)]
    #[replace_float_literals(T::from_f64(literal))]
    fn eval(&self, input: &SVector<T, 1>) -> SVector<T, 1> {
        let vout = input[0];
        let v = T::simd_recip(self.n * self.vt);
        let expin = vout * v;
        if vout.simd_gt(16.0).any() {
            println!();
        }
        let expn = T::simd_exp(expin / self.num_diodes_fwd).simd_min(1e35);
        let expm = T::simd_exp(-expin / self.num_diodes_bwd).simd_min(1e35);
        let res = self.isat * (expn - expm) + 2. * vout - self.vin;
        SVector::<_, 1>::new(res)
    }

    #[cfg_attr(not(test), inline)]
    #[replace_float_literals(T::from_f64(literal))]
    fn j_inv(&self, input: &SVector<T, 1>) -> Option<SMatrix<T, 1, 1>> {
        let vout = input[0];
        let v = T::simd_recip(self.n * self.vt);
        let expin = vout * v;
        let expn = T::simd_exp(expin / self.num_diodes_fwd).simd_min(1e35);
        let expm = T::simd_exp(-expin / self.num_diodes_bwd).simd_min(1e35);
        let res = v * self.isat * (expn / self.num_diodes_fwd + expm / self.num_diodes_bwd) + 2.;
        // Biasing to prevent divisions by zero, less accurate around zero
        let ret = 1e-6.select(res.simd_abs().simd_lt(1e-6), res).simd_recip();
        Some(SMatrix::<_, 1, 1>::new(ret))
    }
}

impl<T: Scalar> DiodeClipper<T> {
    #[replace_float_literals(T::from_f64(literal))]
    pub fn new_silicon(fwd: usize, bwd: usize, vin: T) -> Self {
        Self {
            isat: 4.352e-9,
            n: 1.906,
            vt: 23e-3,
            num_diodes_fwd: T::from_f64(fwd as f64),
            num_diodes_bwd: T::from_f64(bwd as f64),
            vin,
            sim_tol: 1e-3,
            max_iter: 50,
            last_vout: vin.simd_tanh(),
        }
    }

    #[replace_float_literals(T::from_f64(literal))]
    pub fn new_germanium(fwd: usize, bwd: usize, vin: T) -> Self {
        Self {
            isat: 200e-9,
            n: 2.109,
            vt: 23e-3,
            num_diodes_fwd: T::from_f64(fwd as f64),
            num_diodes_bwd: T::from_f64(bwd as f64),
            vin,
            sim_tol: 1e-3,
            max_iter: 50,
            last_vout: vin.simd_tanh(),
        }
    }

    #[replace_float_literals(T::from_f64(literal))]
    pub fn new_led(nf: usize, nb: usize, vin: T) -> DiodeClipper<T> {
        Self {
            isat: 2.96406e-12,
            n: 2.475312,
            vt: 23e-3,
            vin,
            num_diodes_fwd: T::from_f64(nf as f64),
            num_diodes_bwd: T::from_f64(nb as f64),
            sim_tol: 1e-4,
            max_iter: 50,
            last_vout: vin.simd_tanh(),
        }
    }
}

impl<T: Scalar> DSPMeta for DiodeClipper<T> {
    type Sample = T;
}

impl<T: Scalar> DSPProcess<1, 1> for DiodeClipper<T>
where
    T::Element: Float,
{
    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        self.vin = x[0];
        let mut value = SVector::<_, 1>::new(
            self.vin
                .simd_clamp(-self.num_diodes_bwd, self.num_diodes_fwd),
        );
        newton_rhapson_tol_max_iter(self, &mut value, self.sim_tol, self.max_iter);
        self.last_vout = value[0];
        [value[0]]
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct DiodeClipperModel<T> {
    pub a: T,
    pub b: T,
    pub si: T,
    pub so: T,
}

impl<T: Scalar> DSPMeta for DiodeClipperModel<T> {
    type Sample = T;
}

impl<T: Scalar> DiodeClipperModel<T> {
    #[replace_float_literals(T::from_f64(literal))]
    #[inline]
    pub fn eval(&self, x: T) -> T {
        let x = self.si * x;
        let lower = x.simd_lt(-self.a);
        let higher = x.simd_gt(self.b);
        let case1 = -T::simd_ln(1. - x - self.a) - self.a;
        let case2 = T::simd_ln(1. + x - self.b) + self.b;
        case1.select(lower, case2.select(higher, x)) * self.so
    }
}

impl<T: Scalar> Antiderivative<T> for DiodeClipperModel<T> {
    fn evaluate(&self, x: T) -> T {
        self.eval(x) / (self.si * self.so)
    }

    #[replace_float_literals(T::from_f64(literal))]
    fn antiderivative(&self, x: T) -> T {
        let cx = self.si * x;
        let lower = cx.simd_lt(-self.a);
        lower.if_else(
            || {
                let x0 = self.a + cx - 1.0;
                let num = (self.a - 1.0) * cx + x0 * T::simd_ln(-x0);
                let den = self.si * self.si;
                -num / den
            },
            || {
                let higher = cx.simd_gt(self.b);
                higher.if_else(
                    || {
                        let x0 = -self.b + cx + 1.0;
                        let num = (self.b - 1.0) * cx + x0 * T::simd_ln(x0);
                        let den = self.si * self.si;
                        num / den
                    },
                    || x * x / 2.0,
                )
            },
        ) / 2.0
    }
}

impl<T: Scalar> Default for DiodeClipperModel<T> {
    fn default() -> Self {
        Self::new_silicon(1, 1)
    }
}

impl<T: Scalar> DSPProcess<1, 1> for DiodeClipperModel<T> {
    #[inline(always)]
    fn process(&mut self, [x]: [Self::Sample; 1]) -> [Self::Sample; 1] {
        [self.eval(x)]
    }
}

impl<T: Scalar> Saturator<T> for DiodeClipperModel<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        self.eval(x) / (self.si * self.so)
    }
}

impl<T: Scalar, const N: usize> MultiSaturator<T, N> for DiodeClipperModel<T> {
    fn multi_saturate(&self, x: [T; N]) -> [T; N] {
        x.map(|x| self.saturate(x))
    }

    fn update_state_multi(&mut self, _: [T; N], _: [T; N]) {}

    fn sat_jacobian(&self, x: [T; N]) -> [T; N] {
        x.map(|x| self.sat_diff(x))
    }
}

#[cfg(test)]
mod tests {
    use plotters::prelude::*;
    use simba::simd::SimdValue;
    use std::hint;

    use super::{DiodeClipper, DiodeClipperModel};
    use crate::dsp::DSPProcess;
    use crate::util::tests::{Plot, Series};

    fn dc_sweep(name: &str, mut dsp: impl DSPProcess<1, 1, Sample = f32>) {
        let results = Vec::from_iter(
            (-4800..=4800)
                .map(|i| i as f64 / 100.)
                .map(|v| dsp.process([v as f32])[0]),
        );
        let full_name = format!("{name}/dc_sweep");
        let plot_title = format!("DC sweep: {name}");
        Plot {
            title: &plot_title,
            bode: false,
            series: &[Series {
                label: name,
                samplerate: 100.0,
                series: &results,
                color: &Default::default(),
            }],
        }
        .create_svg(format!("plots/saturators/clippers/dc_sweep_{name}.svg"));
        insta::assert_csv_snapshot!(&*full_name, results, { "[]" => insta::rounded_redaction(4) });
    }

    fn drive_test(name: &str, mut dsp: impl DSPProcess<1, 1, Sample = f32>) {
        let sine_it = (0..).map(|i| i as f64 / 10.).map(f64::sin);
        let amp = (0..5000).map(|v| v as f64 / 5000. * 500.);
        let output = sine_it
            .zip(amp)
            .map(|(a, b)| a * b)
            .map(|v| hint::black_box(dsp.process([v as f32])[0]));
        let results = Vec::from_iter(output.map(|v| v.extract(0)));
        let full_name = format!("{name}/drive_test");
        let plot_title = format!("Drive test: {name}");
        Plot {
            title: &plot_title,
            bode: false,
            series: &[Series {
                label: "Output",
                samplerate: 10.,
                series: &results,
                color: &BLUE,
            }],
        }
        .create_svg(format!("plots/saturators/clippers/drive_{name}.svg"));
        insta::assert_csv_snapshot!(&*full_name, results, { "[]" => insta::rounded_redaction(4) });
    }

    #[test]
    fn snapshot_diode_clipper() {
        let clipper = DiodeClipper::new_led(3, 5, 0.0);
        dc_sweep("regressions/clipper_nr", clipper);
        drive_test("regressions/clipper_nr", clipper);
    }

    #[test]
    fn snapshot_diode_clipper_model() {
        let clipper = DiodeClipperModel::<f32>::new_led(3, 5);
        dc_sweep("regressions/clipper_model", clipper);
        drive_test("regressions/clipper_model", clipper);
    }
}
