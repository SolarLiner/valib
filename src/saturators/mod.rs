use crate::dsp::DSP;
use numeric_literals::replace_float_literals;

use crate::saturators::adaa::{Antiderivative, Antiderivative2};
use clippers::DiodeClipperModel;

use crate::Scalar;

pub mod adaa;
pub mod clippers;

#[allow(unused_variables)]
pub trait Saturator<T: Scalar>: Default {
    /// Saturate an input with a frozen state.
    fn saturate(&self, x: T) -> T;

    /// Update the state given an input and the output of `saturate`.
    #[inline(always)]
    fn update_state(&mut self, x: T, y: T) {}

    #[inline(always)]
    #[replace_float_literals(T::from_f64(literal))]
    fn sat_diff(&self, x: T) -> T {
        (self.saturate(x + 1e-4) - self.saturate(x)) / 1e-4
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Linear;

impl<S: Scalar> Saturator<S> for Linear {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x
    }

    #[inline(always)]
    fn sat_diff(&self, _: S) -> S {
        S::one()
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Tanh;

impl<S: Scalar> Saturator<S> for Tanh {
    #[inline(always)]
    fn saturate(&self, x: S) -> S {
        x.simd_tanh()
    }

    #[inline(always)]
    #[replace_float_literals(S::from_f64(literal))]
    fn sat_diff(&self, x: S) -> S {
        1. - x.simd_tanh().simd_powi(2)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Default)]
pub struct Asinh;

impl<T: Scalar> Saturator<T> for Asinh {
    fn saturate(&self, x: T) -> T {
        x.simd_asinh()
    }

    fn sat_diff(&self, x: T) -> T {
        let x0 = x * x + T::one();
        x0.simd_sqrt().simd_recip()
    }
}

#[derive(Debug, Copy, Clone, Default, Eq, PartialEq)]
pub struct Clipper;

impl<S: Scalar> Saturator<S> for Clipper {
    #[inline(always)]
    #[replace_float_literals(S::from_f64(literal))]
    fn saturate(&self, x: S) -> S {
        x.simd_min(1.0).simd_max(-1.0)
    }

    #[inline(always)]
    #[replace_float_literals(S::from_f64(literal))]
    fn sat_diff(&self, x: S) -> S {
        let mask = x.simd_abs().simd_gt(1.0);
        (1.0).select(mask, 0.0)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Blend<T, S> {
    amt: T,
    inner: S,
}

impl<T: Scalar, S: Saturator<T>> Saturator<T> for Blend<T, S> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        x + self.amt * (self.inner.saturate(x) - x)
    }

    #[inline(always)]
    fn update_state(&mut self, x: T, y: T) {
        self.inner.update_state(x, y)
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        T::one() + self.amt * (self.inner.sat_diff(x) - T::one())
    }
}

impl<T: Scalar, S: Default> Default for Blend<T, S> {
    fn default() -> Self {
        Self {
            amt: T::from_f64(0.5),
            inner: S::default(),
        }
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Dynamic<T> {
    Linear,
    Tanh,
    HardClipper,
    DiodeClipper(DiodeClipperModel<T>),
    SoftClipper(Blend<T, DiodeClipperModel<T>>),
}

impl<T: Scalar> Saturator<T> for Dynamic<T> {
    #[inline(always)]
    fn saturate(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.saturate(x),
            Self::HardClipper => Clipper.saturate(x),
            Self::Tanh => Tanh.saturate(x),
            Self::DiodeClipper(clip) => clip.saturate(x),
            Self::SoftClipper(clip) => clip.saturate(x),
        }
    }

    #[inline(always)]
    fn sat_diff(&self, x: T) -> T {
        match self {
            Self::Linear => Linear.sat_diff(x),
            Self::HardClipper => Clipper.sat_diff(x),
            Self::Tanh => Tanh.sat_diff(x),
            Self::DiodeClipper(clip) => clip.sat_diff(x),
            Self::SoftClipper(clip) => clip.sat_diff(x),
        }
    }
}

impl<T> Default for Dynamic<T> {
    fn default() -> Self {
        Self::Linear
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Slew<T> {
    pub max_diff: T,
    last_out: T,
}

impl<T: Scalar> Default for Slew<T> {
    fn default() -> Self {
        Self {
            max_diff: T::from_f64(1.0),
            last_out: T::from_f64(0.0),
        }
    }
}

impl<T: Scalar> DSP<1, 1> for Slew<T> {
    type Sample = T;

    fn latency(&self) -> usize {
        0
    }

    fn reset(&mut self) {
        self.last_out = T::from_f64(0.0);
    }

    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let y = self.slew(x[0]);
        self.last_out = y;
        [y]
    }
}

impl<T: Scalar> Slew<T> {
    pub fn new(samplerate: T, max_diff: T) -> Self {
        Self {
            max_diff: max_diff / samplerate,
            last_out: T::from_f64(0.0),
        }
    }

    pub fn set_max_diff(&mut self, max: T, samplerate: T) {
        self.max_diff = max / samplerate;
    }

    fn slew_diff(&self, x: T) -> T {
        let diff = x - self.last_out;
        diff.simd_clamp(-self.max_diff, self.max_diff)
    }

    fn slew(&self, x: T) -> T {
        self.last_out + self.slew_diff(x)
    }
}

impl<T: Scalar> Saturator<T> for Slew<T> {
    fn saturate(&self, x: T) -> T {
        self.slew(x)
    }

    fn sat_diff(&self, x: T) -> T {
        self.slew_diff(x)
    }

    fn update_state(&mut self, _x: T, y: T) {
        self.last_out = y;
    }
}
