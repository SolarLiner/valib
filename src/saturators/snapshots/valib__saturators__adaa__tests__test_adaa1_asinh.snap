---
source: src/saturators/adaa.rs
expression: "&output as &[_]"
---
0.0
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
-0.75
0.75
1.566
1.771
1.566
0.75
-0.75
-1.566
-1.771
-1.566
